File: nodejs/lib/crypto.js
'use strict'

const crypto = require('crypto')

const AES_CBC_IDENTIFIER = 'aes-128-cbc'
const AES_IV_BYTES = 16
const AES_KEY_BYTES = 16

module.exports = {
  /**
   * Decrypts the given message using the given key and iv.
   *
   * @param {Buffer} message the message to decrypt
   * @param {Buffer} key the key to decrypt the data with
   * @param {Buffer} iv the initialization vector to use
   * @return {string} the decrypt message as string
   * @throws {Error} if decryption fails or input is invalid
   */
  decrypt: (message, key, iv) => {
    // Validate the key and iv
    module.exports.validateKeyAndIv(key, iv)

    // Sanity check the message
    if (!Buffer.isBuffer(message) || message.length === 0) {
      throw new Error('Given message is invalid')
    }

    // Decrypt the message
    const cipher = crypto.createDecipheriv(AES_CBC_IDENTIFIER, key, iv)
    return cipher.update(message, undefined, 'utf8') + cipher.final('utf8')
  },

  /**
   * Encrypts the given message with the given key and initialization vector.
   *
   * @param {Buffer} message the message to encrypt
   * @param {Buffer} key the key to encrypt the data with
   * @param {Buffer} iv the initialization vector to use
   * @return {string} the encrypted message as base64 encoded string
   * @throws {Error} if encryption fails or input is invalid
   */
  encrypt: (message, key, iv) => {
    // Validate the key and iv
    module.exports.validateKeyAndIv(key, iv)

    // Sanity check the message
    if (!Buffer.isBuffer(message) || message.length === 0) {
      throw new Error('Given message is invalid')
    }

    // Encrypt the message
    const cipher = crypto.createCipheriv(AES_CBC_IDENTIFIER, key, iv)
    return cipher.update(message, 'utf-8', 'base64') + cipher.final('base64')
  },

  validateKeyAndIv: (key, iv) => {
    if (!Buffer.isBuffer(key)) {
      throw new TypeError('Encryption key must be a Buffer')
    }

    if (!Buffer.isBuffer(iv)) {
      throw new TypeError('IV must be a Buffer')
    }

    if (key.length !== AES_KEY_BYTES) {
      throw new Error('Encryption key must be ' + AES_KEY_BYTES + ' bytes')
    }

    if (iv.length !== AES_IV_BYTES) {
      throw new Error('IV must be ' + AES_IV_BYTES + ' bytes')
    }

    if (iv.compare(key) === 0) {
      throw new Error('Key and IV cannot be the same!')
    }

    return true
  }
}

File: python/kmscrypt/crypto.py
"""Main module containing encryption and decryption routines."""

import base64
import binascii
import os

from typing import TYPE_CHECKING, Dict, Union

import boto3
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import padding

if TYPE_CHECKING:
    from typing_extensions import Final, TypedDict
else:
    Final = object
    TypedDict = object

AES_IV_BYTES = 16
AES_KEY_BYTES = 16  # 128 bits
AES_BLOCK_SIZE = 128  # bits
AES_KEY_SPEC: Final = "AES_128"


class EncryptedData(TypedDict):
    """Dict containing encrypted data and information needed to decrypt it."""

    EncryptedData: str
    EncryptedDataKey: str
    EncryptionContext: Dict[str, str]
    Iv: str


def decrypt(data: EncryptedData, session: boto3.Session = None) -> bytes:
    """Decrypts previously encrypted data.

    Args:
        data: The (JSON) object returned by earlier call to encrypt().
        session: boto3.Session object to use for KMS calls.

    Returns:
        The encrypted secret as bytes.
    """
    # Decode payload
    ciphertext = base64.b64decode(data["EncryptedData"])
    encrypted_data_key = base64.b64decode(data["EncryptedDataKey"])
    initialization_vector = binascii.unhexlify(data["Iv"])

    # Decrypt key
    kms = (session or boto3.Session()).client("kms")
    res = kms.decrypt(
        CiphertextBlob=encrypted_data_key,
        EncryptionContext=data["EncryptionContext"],
    )
    key = res["Plaintext"]

    # Decrypt data
    cipher = Cipher(algorithms.AES(key), modes.CBC(initialization_vector))
    decryptor = cipher.decryptor()
    padded_plaintext = decryptor.update(ciphertext) + decryptor.finalize()

    # Unpad data
    unpadder = padding.PKCS7(AES_BLOCK_SIZE).unpadder()
    plaintext = unpadder.update(padded_plaintext) + unpadder.finalize()

    return plaintext


def encrypt(
    data: Union[str, bytes],
    key_id: str,
    encryption_context: Dict[str, str] = None,
    session: boto3.Session = None,
) -> EncryptedData:
    """Encrypts a given data string.

    Args:
        data: The secret to encrypt. It can be str / unicode (py2) or
            str / bytes (py3).
        key_id: ID, ARN, alias or alias ARN of the KMS key to encrypt the
            data with
        encryption_context: Optional encryption context (key-value dict)
            for KMS
        session: boto3.Session object to use for KMS calls.

    Returns:
        A dictionary containing the data that is required to decrypt the
        secret.
    """

    if not encryption_context:
        encryption_context = {}

    # Generate key
    kms = (session or boto3.Session()).client("kms")
    res = kms.generate_data_key(
        KeyId=key_id, KeySpec=AES_KEY_SPEC, EncryptionContext=encryption_context
    )
    key = res["Plaintext"]

    # And initialization vector
    initialization_vector = os.urandom(16)

    # Ensure data is bytes
    if isinstance(data, str):
        data = data.encode("utf-8")

    # Pad data
    padder = padding.PKCS7(AES_BLOCK_SIZE).padder()
    padded_data = padder.update(data) + padder.finalize()

    # Encrypt data
    cipher = Cipher(algorithms.AES(key), modes.CBC(initialization_vector))
    encryptor = cipher.encryptor()
    ciphertext = encryptor.update(padded_data) + encryptor.finalize()

    return {
        "EncryptedData": base64.b64encode(ciphertext).decode(),
        "EncryptedDataKey": base64.b64encode(res["CiphertextBlob"]).decode(),
        "EncryptionContext": encryption_context,
        "Iv": binascii.b2a_hex(initialization_vector).decode(),
    }
